internal/pkg/agent/application/upgrade/step_unpack.go

```diff
@@ -7,52 +7,105 @@ package upgrade
+	v1 "github.com/elastic/elastic-agent/pkg/api/v1"
 	"github.com/elastic/elastic-agent/pkg/core/logger"
 )
 
+// UnpackResult contains the location and hash of the unpacked agent files
+type UnpackResult struct {
+	// Hash contains the unpacked agent commit hash, limited to a length of 6 for backward compatibility
+	Hash string `json:"hash" yaml:"hash"`
+	// VersionedHome indicates the path (forward slash separated) where to find the unpacked agent files
+	// The value depends on the mappings specified in manifest.yaml, if no manifest is found it assumes the legacy data/elastic-agent-<hash> format
+	VersionedHome string `json:"versioned-home" yaml:"versioned-home"`
+}
+
 // unpack unpacks archive correctly, skips root (symlink, config...) unpacks data/*
-func (u *Upgrader) unpack(version, archivePath, dataDir string) (string, error) {
+func (u *Upgrader) unpack(version, archivePath, dataDir string) (UnpackResult, error) {
 	// unpack must occur in directory that holds the installation directory
 	// or the extraction will be double nested
-	var hash string
+	var unpackRes UnpackResult
 	var err error
 	if runtime.GOOS == windows {
-		hash, err = unzip(u.log, archivePath, dataDir)
+		unpackRes, err = unzip(u.log, archivePath, dataDir)
 	} else {
-		hash, err = untar(u.log, version, archivePath, dataDir)
+		unpackRes, err = untar(u.log, version, archivePath, dataDir)
 	}
 
 	if err != nil {
-		u.log.Errorw("Failed to unpack upgrade artifact", "error.message", err, "version", version, "file.path", archivePath, "hash", hash)
-		return "", err
+		u.log.Errorw("Failed to unpack upgrade artifact", "error.message", err, "version", version, "file.path", archivePath, "unpackResult", unpackRes)
+		return UnpackResult{}, err
 	}
 
-	u.log.Infow("Unpacked upgrade artifact", "version", version, "file.path", archivePath, "hash", hash)
-	return hash, nil
+	u.log.Infow("Unpacked upgrade artifact", "version", version, "file.path", archivePath, "unpackResult", unpackRes)
+	return unpackRes, nil
 }
 
-func unzip(log *logger.Logger, archivePath, dataDir string) (string, error) {
+func unzip(log *logger.Logger, archivePath, dataDir string) (UnpackResult, error) {
 	var hash, rootDir string
 	r, err := zip.OpenReader(archivePath)
 	if err != nil {
-		return "", err
+		return UnpackResult{}, err
 	}
 	defer r.Close()
 
 	fileNamePrefix := strings.TrimSuffix(filepath.Base(archivePath), ".zip") + "/" // omitting `elastic-agent-{version}-{os}-{arch}/` in filename
 
+	pm := pathMapper{}
+	versionedHome := ""
+
+	// Load manifest, the use of path.Join is intentional since in .zip file paths use slash ('/') as separator
+	manifestFile, err := r.Open(path.Join(fileNamePrefix, "manifest.yaml"))
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		// we got a real error looking up for the manifest
+		return UnpackResult{}, fmt.Errorf("looking up manifest in package: %w", err)
+	}
+	if err == nil {
+		// load manifest
+		defer manifestFile.Close()
```

Comment:
same here would be nice to release manifest handler sooner

Reply:
consider adding helper which parses it and returns structure as this is recurring or pathmapper constructor with filepath as arg

Reply:
Fixed in #4177 
https://github.com/elastic/elastic-agent/pull/4177/files#diff-7f896b3c21ff10574e4b2bfaff289b3ce21f7f34450b815a8a6e838433caa848R252